src/services/spot/Spot.cpp

```diff
@@ -0,0 +1,387 @@
+              doc.AddMember("yAxis",y_axis_value,doc.GetAllocator());
+              doc.AddMember("title",title_value,doc.GetAllocator());
+              doc.AddMember("show_exclusive", show_exclusive, doc.GetAllocator());
+              auto& json_commits = doc["commits"];
+              auto& json_times = doc["XTics"];
+              json_commits.GetArray().PushBack(commit_value,doc.GetAllocator());
+              json_times.GetArray().PushBack(xtic_value,doc.GetAllocator());
+              rapidjson::Value& series_array = doc["series"]; 
+              for(auto datum : json) {
+                 std::string series_name = datum.first;
+                 if(series_name.size()>1){
+                   rapidjson::Value value_series_name; 
+                   value_series_name.SetString(series_name.c_str(),doc.GetAllocator());
+                   series_array.GetArray().PushBack(value_series_name,doc.GetAllocator());
+                   rapidjson::Value outarr;
+                   outarr.SetArray();
+                   rapidjson::Value arrarr;
+                   arrarr.SetArray();
+                   arrarr.PushBack(0,doc.GetAllocator());
+                   arrarr.PushBack(((float)datum.second)/(1.0*divisor),doc.GetAllocator());
+                   outarr.PushBack(arrarr,doc.GetAllocator());
+                   value_series_name.SetString(series_name.c_str(),doc.GetAllocator());
+                   doc.AddMember(value_series_name,outarr,doc.GetAllocator());
+                 }
+              } 
+            }
+            std::ofstream ofs(place.c_str());
+            rapidjson::OStreamWrapper osw(ofs);
+            rapidjson::Writer<rapidjson::OStreamWrapper> writer(osw);
+            doc.Accept(writer);
+          }
+        }
+        // TODO: reimplement
+        Spot()
+            { }
+
+        //
+        // --- callback functions
+        //
+        static std::pair<AggregationHandler,SelectionHandler> create_query_processor(std::string query){
+           CalQLParser parser(query.c_str()); 
+           if (parser.error()) {
+               Log(0).stream() << "spot: config parse error: " << parser.error_msg() << std::endl;
+               return std::make_pair(nullptr,nullptr);
+           }
+           QuerySpec    spec(parser.spec());
+           return std::make_pair(new Aggregator(spec),new RecordSelector(spec));
+        }
+        static std::string query_for_annotation(std::string grouping, std::string metric = "time.inclusive.duration"){
+          std::string end_grouping = "event.end#"+grouping;
+          return "SELECT " + grouping+","+end_grouping+",sum("+metric+") " + "WHERE " +grouping+","+end_grouping+","+metric + " GROUP BY " + end_grouping+","+grouping;
+        }
+        static void pre_write_cb(Caliper* c, const SnapshotRecord* flush_info) {
+            ConfigSet    config(RuntimeConfig::init("spot", s_configdata));
+            const std::string&  config_string = config.get("config").to_string().c_str();
```

Comment:
Should just be `std::string  config_string = config.get("config").to_string();` How did this even compile?